File: lib/ts/src/DefaultSerialization.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { Packr, Unpackr } from 'msgpackr';
import { Serialization } from './Serialization.js';

export class DefaultSerialization implements Serialization {
    private textEncoder = new TextEncoder();
    private textDecoder = new TextDecoder();
    private packr = new Packr({ mapsAsObjects: false, useRecords: false });
    private unpackr = new Unpackr({ mapsAsObjects: false, useRecords: false });

    public toJson(telepactMessage: any): Uint8Array {
        const jsonStr = JSON.stringify(telepactMessage);
        return this.textEncoder.encode(jsonStr);
    }

    public toMsgpack(telepactMessage: any): Uint8Array {
        return this.packr.encode(telepactMessage);
    }

    public fromJson(bytes: Uint8Array): any {
        const jsonStr = this.textDecoder.decode(bytes);
        return JSON.parse(jsonStr);
    }

    public fromMsgpack(bytes: Uint8Array): any {
        return this.unpackr.decode(bytes);
    }
}

File: lib/py/msgpact/DefaultSerialization.py
import msgpack
import json

from .Serialization import Serialization


class DefaultSerialization(Serialization):

    def to_json(self, msgpact_message: object) -> bytes:
        return json.dumps(msgpact_message).encode()

    def to_msgpack(self, msgpact_message: object) -> bytes:
        return msgpack.dumps(msgpact_message)

    def from_json(self, bytes_: bytes) -> object:
        return json.loads(bytes_)

    def from_msgpack(self, bytes_: bytes) -> object:
        return msgpack.loads(bytes_, strict_map_key=False)
